numpy/_core/src/umath/clip.cpp

```diff
@@ -150,50 +152,90 @@ _NPY_CLIP(T x, T min, T max)
+        for (npy_intp i = 0; i < n; i++, ip += is, op += os) {
+            *(T *)op = _NPY_CLIP<Tag>(*(T *)ip, min_val, max_val);
+        }
+    }
+}
 
-        T *ip1 = args[0], *op1 = args[3];
-        npy_intp is1 = steps[0] / sizeof(T), os1 = steps[3] / sizeof(T);
+template <class Tag, class T>
+void
+_npy_clip_const_minmax_(
+    char *ip, npy_intp is, char *op, npy_intp os, npy_intp n, T min_val, T max_val,
+    std::true_type  /* floating point */
+)
+{
+    if (!npy_isnan(min_val) && !npy_isnan(max_val)) {
+        /* min_val and max_val are not nans */
+        /* nans in the input will be propagated naturally */
 
         /* contiguous, branch to let the compiler optimize */
-        if (is1 == 1 && os1 == 1) {
-            for (npy_intp i = 0; i < n; i++, ip1++, op1++) {
-                *op1 = _NPY_CLIP<Tag>(*ip1, min_val, max_val);
+        if (is == sizeof(T) && os == sizeof(T)) {
+            for (npy_intp i = 0; i < n; i++, ip += sizeof(T), op += sizeof(T)) {
+                T x = *(T *)ip;
+                if (x < min_val) x = min_val;
+                if (x > max_val) x = max_val;
+                *(T *)op = x;
             }
         }
         else {
-            for (npy_intp i = 0; i < n; i++, ip1 += is1, op1 += os1) {
-                *op1 = _NPY_CLIP<Tag>(*ip1, min_val, max_val);
+            for (npy_intp i = 0; i < n; i++, ip += is, op += os) {
+                T x = *(T *)ip;
+                if (x < min_val) x = min_val;
+                if (x > max_val) x = max_val;
+                *(T *)op = x;
             }
         }
     }
     else {
-        T *ip1 = args[0], *ip2 = args[1], *ip3 = args[2], *op1 = args[3];
-        npy_intp is1 = steps[0] / sizeof(T), is2 = steps[1] / sizeof(T),
-                 is3 = steps[2] / sizeof(T), os1 = steps[3] / sizeof(T);
-        for (npy_intp i = 0; i < n;
-             i++, ip1 += is1, ip2 += is2, ip3 += is3, op1 += os1)
-            *op1 = _NPY_CLIP<Tag>(*ip1, *ip2, *ip3);
+        /* min_val and/or max_val are nans */
+        T x = npy_isnan(min_val) ? min_val : max_val;
+        for (npy_intp i = 0; i < n; i++, op += os) {
+            *(T *)op = x;
+        }
     }
-    npy_clear_floatstatus_barrier((char *)dimensions);
 }
 
-template <class Tag>
+template <class Tag, class T = typename Tag::type>
 static void
 _npy_clip(char **args, npy_intp const *dimensions, npy_intp const *steps)
 {
-    using T = typename Tag::type;
-    return _npy_clip_<Tag>((T **)args, dimensions, steps);
+    npy_intp n = dimensions[0];
+    if (steps[1] == 0 && steps[2] == 0) {
+        /* min and max are constant throughout the loop, the most common case
+         */
+        /* NOTE: it may be possible to optimize these checks for nan */
+        T min_val = *(T *)args[1];
+        T max_val = *(T *)args[2];
+
+        _npy_clip_const_minmax_<Tag, T>(
+            args[0], steps[0], args[3], steps[3], n, min_val, max_val,
+            std::is_base_of<npy::floating_point_tag, Tag>{}
```

Comment:
I tried this out (with MSVC) and it gave much poorer timings.
It didn't make a difference even if I replaced `steps[0]` and `steps[3]` with `sizeof(T)` in the function call.

Reply:
I assume that was also with the `static inline` already in place?  I am a bit surprised that MSVC doesn't optimize it (and it makes me wonder if there quite a few places in NumPy where MSVC fails to optimize...).

Reply:
Yes, with `static inline` already in place.

Reply:
Makes me really wonder if should just use `NPY_FINLINE` here, but maybe more importantly in the other places that try to use this auto-specialization trick...

Auto-specialization is really handy, but if MSVC thwarts it may be getting slow execution on windows for a few functions for no good reason...